### PR TITLE
Added simple image visualisations

### PR DESCRIPTION
**Pull request by alecokas**
* Added functionality to visualise reconstruction, single latent dimension traversals, complete latent space traversal, and random sample visualisation with a basic CLI.
* Added shuffle parameter to the ```dataloader```s.
* Made some minor variable name changes for clarity.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -166,7 +166,7 @@ def main(args):
                           else "cpu")
 
     # PREPARES DATA
-    train_loader, test = get_dataloaders(batch_size=args.batch_size, dataset=args.dataset)
+    train_loader, test_loader = get_dataloaders(batch_size=args.batch_size, dataset=args.dataset)
```
Comment (linesd) [inline main.py]: I agree this is clearer

File: utils/dataloaders.py [hunk 2]
```diff
@@ -29,11 +29,11 @@ def get_dataloaders(batch_size, dataset):
         "celeba": get_celeba_dataloader
     }
     if dataset in dataset_options:
-        return dataset_options[dataset](batch_size=batch_size)
+        return dataset_options[dataset](batch_size=batch_size, shuffle=shuffle)
```
Comment (linesd) [inline utils/dataloaders.py]: Is it not neater to just call shuffle here rather than shuffle=shuffle?

File: utils/dataloaders.py [hunk 3]
```diff
@@ -29,11 +29,11 @@ def get_dataloaders(batch_size, dataset):
...
     else:
         raise Exception("{} is not valid. Please enter a valid dataset".format(dataset))
 
-def get_mnist_dataloaders(batch_size=128,
+def get_mnist_dataloaders(batch_size=128, shuffle=False,
```
Comment (linesd) [inline utils/dataloaders.py]: Here you are setting shuffle=False in the get_mnist_dataloaders function call, but in the call to DataLoader below you are resetting it to True. So the argument in get_mnist_dataloaders can't affect the call to DataLoader. I'm assuming this was not the intention?

File: run_experiment.py [hunk 4]
```diff
@@ -31,6 +33,17 @@ def parse_arguments():
     experiment.add_argument('-x', '--experiment',
                             default='custom', choices=experiment_options,
                             help='Predefined experiments to run. If not `custom` this will set the correct other arguments.')
+
+    visualisation = parser.add_argument_group('Desired Visualisation')
+    visualisation_options = ['random_samples', 'traverse_all_latent_dims', 'traverse_one_latent_dim', 'random_reconstruction']
+    visualisation.add_argument('-v', '--visualisation',
+                            default='random_samples', choices=visualisation_options,
+                            help='Predefined visaulisation options which can be performed.')
```
Comment (linesd) [inline run_experiment.py]: Small spelling error in visualisation.

File: viz/visualize.py [hunk 5]
```diff
@@ -43,10 +43,10 @@ def reconstructions(self, data, size=(8, 8), filename='recon.png'):
         self.model.train()
 
         # Upper half of plot will contain data, bottom half will contain
-        # reconstructions
-        num_images = size[0] * size[1] / 2
-        originals = input_data[:num_images].cpu()
-        reconstructions = recon_data.view(-1, *self.model.img_size)[:num_images].cpu()
+        # reconstructions of the original image
+        num_images = size[0]
+        originals = input_data.cpu()
```
Comment (linesd) [inline viz/visualize.py]: Do these always get sent to cpu? Does it not conflict with `input_data = data.to(self.model.device)` where the `model.device` determines where it is sent? I'm familiar enough with PyTorch to know this.

File: utils/dataloaders.py (hunk 2 above)
Comment (alecokas) [reply utils/dataloaders.py]: I like to be explicit about which parameter is being called, but I don't mind changing it if it is prefered @YannDubs .

File: utils/dataloaders.py (hunk 3 above)
Comment (alecokas) [reply utils/dataloaders.py]: This just sets the default to ```False```. Anything passed to the function will override the default ```shuffle=False``` setting.

File: viz/visualize.py (hunk 5 above)
Comment (alecokas) [reply viz/visualize.py]: Seems so. I left the ```.cpu()``` call operating in the way that I found it.

File: run_experiment.py (hunk 4 above)
Comment (alecokas) [reply run_experiment.py]: Nice spot. Thanks.

File: utils/dataloaders.py (hunk 3 above)
Comment (linesd) [reply utils/dataloaders.py]: Yes, I'm aware it's just the default setting. But inside the function `get_mnist_dataloaders` you are calling the function `DataLoader` of which one of the arguments is `shuffle=True`. My question is, should the the internal function `DataLoader` always have the argument  `shuffle=True`? Or should it inherit the boolean value of the parent function?

File: utils/dataloaders.py (hunk 2 above)
Comment (linesd) [reply utils/dataloaders.py]: Happy to leave it if you would rather be explicit about it.

File: utils/dataloaders.py (hunk 3 above)
Comment (alecokas) [reply utils/dataloaders.py]: Oh I understand. Yeah I think that you are correct. I'll make the required change. Thanks.

Comment (linesd) [reply utils/dataloaders.py]: Cool. FYI -  this error occurs in a number of the functions in this file.

File: run_experiment.py [hunk 6]
```diff
@@ -8,15 +8,17 @@
...
+def random_samples(experiment_name, num_samples=1):
+    """ generate a number of random samples from the dataset
     """
     with open('trained_models/{}/specs.json'.format(experiment_name)) as spec_file:
         spec_data = json.load(spec_file)
         dataset_name = spec_data['dataset']
 
-    _, test = get_dataloaders(batch_size=batch_size, dataset=dataset_name)
-    return test
+        _, data_loader = get_dataloaders(batch_size=num_samples, dataset=dataset_name, shuffle=True)
```
Comment (YannDubs) [inline run_experiment.py]: don't you want keep test_loader here to know what data we are talking about ?

File: utils/dataloaders.py (hunk 2 above)
Comment (YannDubs) [reply utils/dataloaders.py]: I normally always give the argument key and value (i.e `shuffle=shuffle`) if it's an optional argument (to avoid possible bugs  / not have to remember the order in which I wrote the function / works on all python version which is not the case if you don't) but never for required arguments (to be more concise). I.e. I would not for the `batch_size` in this case.

File: viz/visualize.py (hunk 5 above)
Comment (YannDubs) [reply viz/visualize.py]: @linesd @alecokas when you train or test the model you want all to be on your most powerful device to make it quicker, but when you save it has to be on CPU such that anyone can read it! Good  question though :)

**Comment by YannDubs** [comment]
Added a few comments : but feel free to merge when you're @alecokas  happy (and changed the shuffle thing).

Good review @linesd  💯

File: run_experiment.py (hunk 6 above)
Comment (alecokas) [reply run_experiment.py]: Sure, I'll make this clear.

**Comment by alecokas** [comment]
Merging after resolving the suggested corrections
